testcases/kernel/syscalls/setpgrp/setpgrp02.c

```diff
@@ -1,130 +1,41 @@
- *   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
- */
+// SPDX-License-Identifier: GPL-2.0-or-later
 
 /*
- * NAME
- *	setpgrp02.c
- *
- * DESCRIPTION
- *	Testcase to check the basic functionality of the setpgrp(2) syscall.
- *
- * ALGORITHM
- *	Check the values that setpgrp() and getpgrp() return. The setpgrp()
- *	returns 0 on success in Linux, but, in DYNIX/ptx this call returns
- *	the new pgid.
- *
- * USAGE:  <for command-line>
- *  setpgrp02 [-c n] [-f] [-i n] [-I x] [-P x] [-t]
- *     where,  -c n : Run n copies concurrently.
- *             -f   : Turn off functionality Testing.
- *             -i n : Execute test n times.
- *             -I x : Execute test for x seconds.
- *             -P x : Pause for x seconds between iterations.
- *             -t   : Turn on syscall timing.
- *
- * HISTORY
+ * Copyright (c) International Business Machines  Corp., 2001
  *	07/2001 Ported by Wayne Boyer
- *
- * RESTRICTIONS
- *	None
+ * Copyright (C) 2024 SUSE LLC Andrea Manzini <andrea.manzini@suse.com>
  */
-#include <errno.h>
-#include <sys/wait.h>
-#include "test.h"
 
-char *TCID = "setpgrp02";
-int TST_TOTAL = 1;
+/*\
+ * [Description]
+ * Testcase to check the basic functionality of the setpgrp(2) syscall.
+ */
 
-void setup(void);
-void cleanup(void);
+#include "tst_test.h"
 
-int main(int ac, char **av)
+static void verify_setpgrp(void)
 {
-	int lc;
-
-	int pid, oldpgrp;
-	int e_code, status, retval = 0;
-
-	tst_parse_opts(ac, av, NULL, NULL);
-
-	setup();
-
-	for (lc = 0; TEST_LOOPING(lc); lc++) {
-
-		/* reset tst_count in case we are looping */
-		tst_count = 0;
-
-		if ((pid = FORK_OR_VFORK()) == -1) {
-			tst_brkm(TBROK, cleanup, "fork() failed");
-		}
-
-		if (pid == 0) {	/* child */
-			oldpgrp = getpgrp();
-
-			TEST(setpgrp());
-
-			if (TEST_RETURN != 0) {
-				retval = 1;
-				tst_resm(TFAIL, "setpgrp() FAILED, errno:%d",
-					 errno);
-				continue;
-			}
-
-			if (getpgrp() == oldpgrp) {
-				retval = 1;
-				tst_resm(TFAIL, "setpgrp() FAILED to set "
-					 "new group id");
-				continue;
-			} else {
-				tst_resm(TPASS, "functionality is correct");
-			}
-			exit(retval);
-		} else {	/* parent */
-			/* wait for the child to finish */
-			wait(&status);
-			/* make sure the child returned a good exit status */
-			e_code = status >> 8;
-			if ((e_code != 0) || (retval != 0)) {
-				tst_resm(TFAIL, "Failures reported above");
-			}
-			cleanup();
-		}
+	/* fork() is needed in order to make sure to have a different PGID */
+	int pid = SAFE_FORK();
+
+	if (pid == 0) { // child
+		int oldpgrp = getpgrp();
+
+		TST_EXP_PASS(setpgrp());
+		if (getpgrp() == oldpgrp)
+			tst_res(TFAIL, "setpgrp() FAILED to set new group id");
+		else
+			tst_res(TPASS, "functionality is correct");
+	} else { // parent
+		int status;
+
+		SAFE_WAIT(&status);
+		if (!WIFEXITED(status) || WEXITSTATUS(status) != 0)
+			tst_res(TFAIL, "child 1: %s", tst_strstatus(status));
```

Comment:
We don't need `else` part at all, please remove it (LTP library does the check itself, try to `exit(1)` in child.
Also, if both parent and child were needed and code is longer, it's better to handle them with single `if` for child which exits:

```C
if (!SAFE_FORK()) {
    int oldpgrp = getpgrp();
    TST_EXP_PASS(setpgrp());
    exit(TST_RET);
}

... // some parent code